File: it/test-interfaces/expected/suite/java/test/Entry.java
package test;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;

@JsonTypeInfo(use=JsonTypeInfo.Id.NAME, include=JsonTypeInfo.As.PROPERTY, property="type")
@JsonSubTypes({
  @JsonSubTypes.Type(name="foo", value=Entry.A.class),
  @JsonSubTypes.Type(name="b", value=Entry.B.class),
  @JsonSubTypes.Type(name="Bar", value=Entry.Bar.class),
  @JsonSubTypes.Type(name="Baz", value=Entry.Baz.class)
})
public interface Entry {
  public static class A implements Entry {
    @JsonCreator
    public A() {
    }

    @Override
    public int hashCode() {
      int result = 1;
      return result;
    }

    @Override
    public boolean equals(final Object other) {
      if (other == null) {
        return false;
      }

      if (!(other instanceof A)) {
        return false;
      }

      @SuppressWarnings("unchecked")
      final A o = (A) other;

      return true;
    }

    @Override
    public String toString() {
      final StringBuilder b = new StringBuilder();

      b.append("A");
      b.append("(");
      b.append(")");

      return b.toString();
    }

    public static class Builder {
      public A build() {

        return new A();
      }
    }
  }

  public static class B implements Entry {
    @JsonCreator
    public B() {
    }

    @Override
    public int hashCode() {
      int result = 1;
      return result;
    }

    @Override
    public boolean equals(final Object other) {
      if (other == null) {
        return false;
      }

      if (!(other instanceof B)) {
        return false;
      }

      @SuppressWarnings("unchecked")
      final B o = (B) other;

      return true;
    }

    @Override
    public String toString() {
      final StringBuilder b = new StringBuilder();

      b.append("B");
      b.append("(");
      b.append(")");

      return b.toString();
    }

    public static class Builder {
      public B build() {

        return new B();
      }
    }
  }

  public static class Bar implements Entry {
    @JsonCreator
    public Bar() {
    }

    @Override
    public int hashCode() {
      int result = 1;
      return result;
    }

    @Override
    public boolean equals(final Object other) {
      if (other == null) {
        return false;
      }

      if (!(other instanceof Bar)) {
        return false;
      }

      @SuppressWarnings("unchecked")
      final Bar o = (Bar) other;

      return true;
    }

    @Override
    public String toString() {
      final StringBuilder b = new StringBuilder();

      b.append("Bar");
      b.append("(");
      b.append(")");

      return b.toString();
    }

    public static class Builder {
      public Bar build() {

        return new Bar();
      }
    }
  }

  public static class Baz implements Entry {
    @JsonCreator
    public Baz() {
    }

    @Override
    public int hashCode() {
      int result = 1;
      return result;
    }

    @Override
    public boolean equals(final Object other) {
      if (other == null) {
        return false;
      }

      if (!(other instanceof Baz)) {
        return false;
      }

      @SuppressWarnings("unchecked")
      final Baz o = (Baz) other;

      return true;
    }

    @Override
    public String toString() {
      final StringBuilder b = new StringBuilder();

      b.append("Baz");
      b.append("(");
      b.append(")");

      return b.toString();
    }

    public static class Builder {
      public Baz build() {

        return new Baz();
      }
    }
  }
}

File: it/test-interfaces/expected/suite/csharp/Test/Entry.cs
using JsonSubTypes;
using Newtonsoft.Json;
using System;
using System.Text;

namespace Test {
  [JsonConverter(typeof(JsonSubtypes), "@type")]
  [JsonSubtypes.KnownSubType(typeof(Entry.A), "foo")]
  [JsonSubtypes.KnownSubType(typeof(Entry.B), "b")]
  [JsonSubtypes.KnownSubType(typeof(Entry.Bar), "Bar")]
  [JsonSubtypes.KnownSubType(typeof(Entry.Baz), "Baz")]
  public abstract class Entry {
    [JsonProperty("@type", Required = Required.DisallowNull)]
    private String TypeField {
      get;
    }

    public Entry(
      String TypeField
    ) {
      this.TypeField = TypeField;
    }

    [JsonObject(ItemNullValueHandling = NullValueHandling.Ignore)]
    public class A : Entry {
      [JsonProperty("shared", Required = Required.DisallowNull)]
      public String shared {
        get;
      }

      [JsonConstructor]
      public A(
        [JsonProperty("@type", Required = Required.DisallowNull)] String TypeField,
        [JsonProperty("shared", Required = Required.DisallowNull)] String shared
      ) : base(TypeField) {
        this.shared = shared;
      }

      public override Int32 GetHashCode() {
        Int32 result = 1;
        result = result * 31 + this.shared.GetHashCode();
        return result;
      }

      public override Boolean Equals(Object other) {
        A o = other as A;

        if (o == null) {
          return false;
        }

        if (!this.shared.Equals(o.shared)) {
          return false;
        }

        return true;
      }

      public override String ToString() {
        StringBuilder b = new StringBuilder();

        b.Append("A");
        b.Append("(");
        b.Append("shared=");
        b.Append(this.shared);
        b.Append(")");

        return b.ToString();
      }
    }

    [JsonObject(ItemNullValueHandling = NullValueHandling.Ignore)]
    public class B : Entry {
      [JsonProperty("shared", Required = Required.DisallowNull)]
      public String shared {
        get;
      }

      [JsonConstructor]
      public B(
        [JsonProperty("@type", Required = Required.DisallowNull)] String TypeField,
        [JsonProperty("shared", Required = Required.DisallowNull)] String shared
      ) : base(TypeField) {
        this.shared = shared;
      }

      public override Int32 GetHashCode() {
        Int32 result = 1;
        result = result * 31 + this.shared.GetHashCode();
        return result;
      }

      public override Boolean Equals(Object other) {
        B o = other as B;

        if (o == null) {
          return false;
        }

        if (!this.shared.Equals(o.shared)) {
          return false;
        }

        return true;
      }

      public override String ToString() {
        StringBuilder b = new StringBuilder();

        b.Append("B");
        b.Append("(");
        b.Append("shared=");
        b.Append(this.shared);
        b.Append(")");

        return b.ToString();
      }
    }

    [JsonObject(ItemNullValueHandling = NullValueHandling.Ignore)]
    public class Bar : Entry {
      [JsonProperty("shared", Required = Required.DisallowNull)]
      public String shared {
        get;
      }

      [JsonConstructor]
      public Bar(
        [JsonProperty("@type", Required = Required.DisallowNull)] String TypeField,
        [JsonProperty("shared", Required = Required.DisallowNull)] String shared
      ) : base(TypeField) {
        this.shared = shared;
      }

      public override Int32 GetHashCode() {
        Int32 result = 1;
        result = result * 31 + this.shared.GetHashCode();
        return result;
      }

      public override Boolean Equals(Object other) {
        Bar o = other as Bar;

        if (o == null) {
          return false;
        }

        if (!this.shared.Equals(o.shared)) {
          return false;
        }

        return true;
      }

      public override String ToString() {
        StringBuilder b = new StringBuilder();

        b.Append("Bar");
        b.Append("(");
        b.Append("shared=");
        b.Append(this.shared);
        b.Append(")");

        return b.ToString();
      }
    }

    [JsonObject(ItemNullValueHandling = NullValueHandling.Ignore)]
    public class Baz : Entry {
      [JsonProperty("shared", Required = Required.DisallowNull)]
      public String shared {
        get;
      }

      [JsonConstructor]
      public Baz(
        [JsonProperty("@type", Required = Required.DisallowNull)] String TypeField,
        [JsonProperty("shared", Required = Required.DisallowNull)] String shared
      ) : base(TypeField) {
        this.shared = shared;
      }

      public override Int32 GetHashCode() {
        Int32 result = 1;
        result = result * 31 + this.shared.GetHashCode();
        return result;
      }

      public override Boolean Equals(Object other) {
        Baz o = other as Baz;

        if (o == null) {
          return false;
        }

        if (!this.shared.Equals(o.shared)) {
          return false;
        }

        return true;
      }

      public override String ToString() {
        StringBuilder b = new StringBuilder();

        b.Append("Baz");
        b.Append("(");
        b.Append("shared=");
        b.Append(this.shared);
        b.Append(")");

        return b.ToString();
      }
    }
  }
}
